subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/IncrementalAnnotationProcessingCompileTask.java

```diff
@@ -0,0 +1,122 @@
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+
+/**
+ * Wraps another {@link JavaCompiler.CompilationTask} and sets up its annotation processors
+ * according to the provided processor declarations and processor path. Incremental processors
+ * are decorated in order to validate their behavior.
+ */
+class IncrementalAnnotationProcessingCompileTask implements JavaCompiler.CompilationTask {
+
+    private final JavaCompiler.CompilationTask delegate;
+    private final Set<AnnotationProcessorDeclaration> processorDeclarations;
+    private final List<File> annotationProcessorPath;
+
+    private URLClassLoader processorClassloader;
+    private boolean called;
+
+    IncrementalAnnotationProcessingCompileTask(JavaCompiler.CompilationTask delegate, Set<AnnotationProcessorDeclaration> processorDeclarations, List<File> annotationProcessorPath) {
+        this.delegate = delegate;
+        this.processorDeclarations = processorDeclarations;
+        this.annotationProcessorPath = annotationProcessorPath;
+    }
+
+    @Override
+    public void setProcessors(Iterable<? extends Processor> processors) {
+        throw new UnsupportedOperationException("This decorator already handles annotation processing");
+    }
+
+    @Override
+    public void setLocale(Locale locale) {
+        delegate.setLocale(locale);
+    }
+
+    @Override
+    public Boolean call() {
+        if (called) {
+            throw new IllegalStateException("Cannot reuse a compilation task");
+        }
+        called = true;
+        try {
+            setupProcessors();
+            return delegate.call();
+        } finally {
+            cleanupProcessors();
+        }
+    }
+
+    private void setupProcessors() {
+        processorClassloader = new URLClassLoader(getUrls(annotationProcessorPath));
+        List<Processor> processors = new ArrayList<Processor>(processorDeclarations.size());
+        for (AnnotationProcessorDeclaration declaredProcessor : processorDeclarations) {
+            try {
+                Class<?> processorClass = processorClassloader.loadClass(declaredProcessor.getClassName());
+                Processor processor = (Processor) processorClass.newInstance();
+                processor = makeIncremental(processor, declaredProcessor.getType());
+                processors.add(processor);
+            } catch (Exception e) {
+                throw new IllegalArgumentException(e);
+            }
+        }
+        delegate.setProcessors(processors);
+    }
+
+    private Processor makeIncremental(Processor processor, IncrementalAnnotationProcessorType type) {
+        switch (type) {
+            case SINGLE_ORIGIN:
+                return new SingleOriginProcessor(processor);
+            case MULTIPLE_ORIGIN:
+                return new MultipleOriginProcessor(processor);
+            default:
+                return processor;
+        }
+    }
+
+    private URL[] getUrls(List<File> classpath) {
```

Comment:
:x: should extract this to avoid duplication. We have code in `DefaultClassPath` which does exactly the same thing.

Reply:
Done